pkg/typesystem/typesystem.go

```diff
@@ -0,0 +1,251 @@
+			return true
+		}
+		seenTypes[td.GetType()] = struct{}{}
+	}
+
+	return false
+}
+
+func areUsersetRewritesValid(tds []*openfgapb.TypeDefinition) error {
+	allRelations := map[string]struct{}{}
+	typeToRelations := map[string]map[string]struct{}{}
+	for _, td := range tds {
+		typeName := td.GetType()
+		typeToRelations[typeName] = map[string]struct{}{}
+		for relationName := range td.GetRelations() {
+			typeToRelations[typeName][relationName] = struct{}{}
+			allRelations[relationName] = struct{}{}
+		}
+	}
+
+	for _, td := range tds {
+		for relationName, usersetRewrite := range td.GetRelations() {
+			err := isUsersetRewriteValid(allRelations, typeToRelations[td.GetType()], td.GetType(), relationName, usersetRewrite)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// isUsersetRewriteValid checks if a particular userset rewrite is valid. The first argument is all the relations in
+// the model, the second argument is the subset of relations on the type where the rewrite occurs.
+func isUsersetRewriteValid(allRelations map[string]struct{}, relationsOnType map[string]struct{}, objectType, relation string, usersetRewrite *openfgapb.Userset) error {
+	if usersetRewrite.GetUserset() == nil {
+		return serverErrors.EmptyRewrites(objectType, relation)
+	}
+
+	switch t := usersetRewrite.GetUserset().(type) {
+	case *openfgapb.Userset_ComputedUserset:
+		computedUserset := t.ComputedUserset.GetRelation()
+		if computedUserset == relation {
+			return serverErrors.CannotAllowMultipleReferencesToOneRelation
+		}
+		if _, ok := relationsOnType[computedUserset]; !ok {
+			return serverErrors.RelationNotFound(computedUserset, objectType, nil)
+		}
+	case *openfgapb.Userset_TupleToUserset:
+		tupleset := t.TupleToUserset.GetTupleset().GetRelation()
+		if _, ok := relationsOnType[tupleset]; !ok {
+			return serverErrors.RelationNotFound(tupleset, objectType, nil)
+		}
+
+		computedUserset := t.TupleToUserset.GetComputedUserset().GetRelation()
+		if _, ok := allRelations[computedUserset]; !ok {
+			return serverErrors.RelationNotFound(computedUserset, "", nil)
+		}
+	case *openfgapb.Userset_Union:
+		for _, child := range t.Union.GetChild() {
+			err := isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, child)
+			if err != nil {
+				return err
+			}
+		}
+	case *openfgapb.Userset_Intersection:
+		for _, child := range t.Intersection.GetChild() {
+			err := isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, child)
+			if err != nil {
+				return err
+			}
+		}
+	case *openfgapb.Userset_Difference:
+		err := isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, t.Difference.Base)
+		if err != nil {
+			return err
+		}
+
+		err = isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, t.Difference.Subtract)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func areRelationalTypesValid(tds []*openfgapb.TypeDefinition) error {
```

Comment:
Can we please refactor this so that it's easier to understand and read?

I think it makes more sense for this code to be composed like so:

```
func (t *Typesystem) GetRelation(
    objectType, relation string,
) (*openfgapb.Relation, error) {
    // todo: implement me
}

func (t *Typesystem) DirectlyRelatedUserTypes(
    objectType, relation string,
) ([]*openfgapb.RelationReference, error) {
	// todo: implement me
}

func (t *Typesystem) IsDirectlyAssignable(relation *openfgapb.Relation) (bool, error) {
    return containsSelf(relation.GetRewrite())
}

func containsSelf(rewrite *openfgapb.Userset) bool {
    // same code as isAssignable in this PR
}

func (t *Typesystem) ValidateRelationTypeRestrictions(objectType, relation string) error {
    relation, err := t.GetRelation(objectType, relation)
    if err != nil {
        // return an error because objectType or relation is undefined
    }

    relatedTypes := relation.GetTypeInfo().GetDirectlyRelatedUserTypes()

    assignable := t.IsDirectlyAssignable(relation)
    if assignable && len(relatedTypes) < 1 {
        // return an error since there were no type restrictions specified
    }

    if !assignable && len(relatedTypes) != 0 {
        // return an error since there were type restrictions on an unassignable type
    }

    for _, relatedType := range relatedTypes {
        relatedObjType := relatedType.GetObjectType()
        relatedRelation := relatedType.GetRelation()
        
        if relatedRelation != "" {
            _, err := t.GetRelation(relatedObjType, relatedRelation))
            if err != nil {
                // related object type or relation does not exist, return appropriate err
            }

            continue
        }

        _, err := t.GetTypeDefinition(relatedObjType)
        if err != nil {
            // related type does not exist, return appropriate err
        }		
    }

    return nil
}
```

The way it is written now is quite hard to read because you have to cut through so much boilerplate map code to understand the primary important pieces.

Reply:
Related to https://github.com/openfga/api/pull/33 as well. Makes this composition possible.